signac/synced_collections/data_types/synced_list.py

```diff
@@ -0,0 +1,249 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+"""Implements the :class:`SyncedList`.
+
+This implements a list-like data structure that also conforms to the
+:class:`~.SyncedCollection` API and can be combined with any backend type to
+give a list-like API to a synchronized data structure.
+"""
+
+from collections.abc import MutableSequence, Sequence
+
+from ..utils import AbstractTypeResolver
+from .synced_collection import NUMPY, SyncedCollection, _sc_resolver
+
+if NUMPY:
+    import numpy
+
+
+# Identifies sequences, which are the base type for this class.
+_sequence_resolver = AbstractTypeResolver(
+    {
+        "SEQUENCE": (
+            lambda obj: (isinstance(obj, Sequence) and not isinstance(obj, str))
+            or (NUMPY and (isinstance(obj, numpy.ndarray) and obj.shape != ()))
+        ),
+    }
+)
+
+
+class SyncedList(SyncedCollection, MutableSequence):
+    """Implementation of list data structure.
+
+    The SyncedList inherits from :class:`~synced_collection.SyncedCollection`
+    and :class:`~collections.abc.MutableSequence`. Therefore, it behaves similar
+    to a :class:`list`.
+
+    Warnings
+    --------
+    While the SyncedList object behaves like a :class:`list`, there are
+    important distinctions to remember. In particular, because operations
+    are reflected as changes to an underlying backend, copying (even deep
+    copying) a SyncedList instance may exhibit unexpected behavior. If a
+    true copy is required, you should use the `_to_base()` method to get a
+    :class:`list` representation, and if necessary construct a new
+    SyncedList.
+
+    """
+
+    def __init__(self, data=None, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if data is None:
+            self._data = []
+        else:
+            self._validate(data)
+            if NUMPY and isinstance(data, numpy.ndarray):
+                data = data.tolist()
```

Comment:
For 0d numpy arrays `data.item()` and `data.tolist()` are equivalent (they return Python scalars if an appropriate conversion exists, otherwise a numpy scalar). For 1-element numpy arrays, they behave differently because `tolist()` will preserve the list while `item` will squeeze out all singleton dimensions. If a user manually constructs a `SyncedList` with a scalar or a 0d array, that will fail as expected because it is not a sequence. A nested assignment of a numpy scalar to a SyncedCollection will never trigger this path, because 0d numpy arrays are not identified as one of the base types for SyncedList (see the `_sequence_resolver` above).

#503 helps isolate this logic, hopefully making it clearer in the future.